### PR TITLE
Add named groups for python

### LINKED ISSUE #206: Feature: named capture groups
I've been looking at how cucumber-expressions could be adopted in Behat. 

One feature of our existing two pattern syntaxes (regex and turnip) is that they can name the arguments.

e.g.
```
@When /^I eat (?<count>.*) (?<fruit>.*)$/
@When I eat :count :fruit
```

This allows Behat to match arguments based on name as well as position, letting users transpose the argument order if necessary. 

That can be useful if multiple patterns are attached to one step, something allowed in Behat:
```php
/**
 * @When I eat :count :fruit
 * @When :count of the :fruit are eaten
 * @When I don't eat any :fruit
 */
public function myStepDef(string $fruit, int $count=0): void
```

The new feature would be for Cucumber Expressions to capture argument names:

 1. Define a syntax for adding names to expressions (e.g. `{fruit:string}`)
 2. Retain that name in the generated regex as a named group
 3. Attach the name to the matched Argument value objects that the parser outputs (with an accessor method)

Then it would be up to the Cucumber implementation to either use the names for matching to the step definition, or to ignore them and use the order directly as currently happens.

### PR DESCRIPTION
**Pull request by jsa34**
### 🤔 What's changed?

Expression matching now returns a tuple: the value as before and an optional name if the expression has a name or the regex uses a named capture group.

The format for the Cucumber Expression when specifying name AND type is:

```
"There are {step_count:int} steps."
```
Where the part before the colon is the name of the arg, and after is the type we are currently using for Cucumber Expressions.

**The return type is change is currently a breaking change, and will have obviously need to put it probably behind some feature flag, and make the default return the old, expected single value again, and the new tuple only if enabled.  I have not done this yet as I wanted to check the breadth of test cases if the feature was fully enabled for the PoC before implementing. Also, not sure how best to feature flag!**

To resolve #206 

### ⚡️ What's your motivation? 

Python (in particular pytest-bdd) uses other args in the step definitions, such as fixtures and reserved args for "datatable" and "docstring", so just mapping step arg values to step args in the expressions is not reliable or user-friendly.  It is a blocker currently for adopting Cucumber Expressions into the pytest-bdd framework.

### 🏷️ What kind of change is this?

- :boom: Breaking change (incompatible changes to the API) 

### ♻️ Anything particular you want feedback on?

Approach, public API changes, format of the named args, whether it's acceptable in general!

### 📋 Checklist:

- [x] I agree to respect and uphold the [Cucumber Community Code of Conduct](https://github.com/cucumber/.github/tree/main?tab=coc-ov-file)
- [X] I've changed the behaviour of the code
  - [ ] I have added/updated tests to cover my changes.
- [X] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.
- [X] Users should know about my change
  - [ ] I have added an entry to the "Unreleased" section of the [**CHANGELOG**](../blob/main/CHANGELOG.md), linking to this pull request.

----

*This text was originally generated from a [template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/about-issue-and-pull-request-templates), then edited by hand. [You can modify the template here.](https://github.com/cucumber/.github/edit/main/.github/PULL_REQUEST_TEMPLATE.md)*

### CODE REVIEW

**Comment by jsa34** [comment]
This is just my first draft where I made sure that everything "works", but I still need to do tests, documentation, etc., if we agree, but I thought it more prudent to share the current approach I am suggesting before doing these things!

Feedback most welcome!

**Comment by luke-hill** [comment]
The last time we've done a big change like this I think was the abolition of the Transformer proc. And the introduction of this "proper". So I'd prefer that we go down the route of releasing this all simultaneously.

In terms of feature flags, I'm happy for it to sit behind feature flags - but I'd rather it be simple and purely from a dev POV to avoid us needing to review a leviathan PR.

I'm only one of the main contributors though, so it's not entirely my decision. But when cucumber expressions were released, they were released simultaneously - Admittedly this was a long time ago with many more different people at the helm.

TL;DR - I'm pro this change, and anti it sitting behind feature flags when released (But during dev go for it).

**Review by mpkorstanje — CHANGES_REQUESTED** [review]
Mostly looks good to me, but I do have some remarks, nothing major though. Apologies if they're a bit scattered, I'm throwing these out as I go through the PR.

Also please do add an entry to the CHANGELOG.

File: python/tests/test_expression.py [hunk 1]
```diff
@@ -12,18 +13,14 @@
...
-    ]
+    return [yaml_dir / file for file in yaml_dir.iterdir() if file.suffix == ".yaml"]
 
 
 def match(
     expression: str,
     match_text: str,
     parameter_registry: ParameterTypeRegistry = ParameterTypeRegistry(),
-):
+) -> Optional[Tuple[Any, str]]:
```
Comment (mpkorstanje) [inline python/tests/test_expression.py:23]: In other languages a linked map would be preferable I think.

File: python/cucumber_expressions/argument.py [hunk 2]
```diff
@@ -1,42 +1,52 @@
...
-        tree_regexp: TreeRegexp, text: str, parameter_types: List
+        tree_regexp: TreeRegexp,
+        text: str,
+        parameter_types_and_names: List[Tuple[ParameterType, Optional[str]]],
     ) -> Optional[List[Argument]]:
+        # Check if all elements in parameter_types_and_names are tuples
+        for item in parameter_types_and_names:
+            if not isinstance(item, tuple) or len(item) != 2:
+                raise CucumberExpressionError(
+                    f"Expected a tuple of (ParameterType, Optional[str]), but got {type(item)}: {item}"
```
Comment (mpkorstanje) [inline python/cucumber_expressions/argument.py:28]: This error is very technical. What should a user do if/when they encounter this error?

File: python/cucumber_expressions/expression.py [hunk 3]
```diff
@@ -87,20 +86,34 @@ def rewrite_alternation(self, node: Node):
...
         if len(regexps) == 1:
             return rf"({regexps[0]})"
         return rf"((?:{')|(?:'.join(regexps)}))"
 
+    def parse_parameter_name(
+        self, name: str
+    ) -> Tuple[Optional[str], Optional[ParameterType]]:
+        """Helper function to parse the parameter name and return group_name and parameter_type."""
+        if ":" in name:
+            group_name, parameter_type_name = [part.strip() for part in name.split(":")]
```
Comment (mpkorstanje) [inline python/cucumber_expressions/expression.py:108]: This allows for the empty group name, which is distinct from the `None` group name. Probably now what we want.

File: python/cucumber_expressions/expression_factory.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+    @staticmethod
+    def _has_curly_brackets(string: str) -> bool:
+        return "{" in string and "}" in string
+
+    @staticmethod
+    def _extract_text_in_curly_brackets(string: str) -> list:
+        return CURLY_BRACKET_PATTERN.findall(string)
+
+    def is_cucumber_expression(self, expression_string: str):
```
Comment (mpkorstanje) [inline python/cucumber_expressions/expression_factory.py:25]: I don't think this check is simple enough. The primary constraint is explaining to people what is and is not a cucumber expression. For Java I eventually settled on requiring that all regular expressions start with `^` or end with `$` and that everything else is a Cucumber expression. This is both simple and unambiguous.

File: python/cucumber_expressions/tree_regexp.py [hunk 5]
```diff
@@ -51,22 +57,30 @@ def create_group_builder(self, regexp):
...
+        """
+        group_name_start = index + 3
+        group_name_end = source.find(">", group_name_start)
+        return source[group_name_start:group_name_end]
+
+    @staticmethod
+    def is_non_capturing(source: str, index: int) -> bool:
+        # Check if it's a non-capturing group like (?:...)
         if source[index + 1] != "?":
-            # (X)
```
Comment (mpkorstanje) [inline python/cucumber_expressions/tree_regexp.py:57]: Please put these comments back. It's really helpful to have a reference here.

File: python/cucumber_expressions/tree_regexp.py [hunk 6]
```diff
@@ -51,22 +57,30 @@ def create_group_builder(self, regexp):
...
-        if source[index + 2] != "<":
-            # (?:X)
-            # (?idmsuxU-idmsuxU)
-            # (?idmsux-idmsux:X)
-            # (?=X)
-            # (?!X)
-            # (?>X)
-            return True
-        # (?<=X) or (?<!X) else (?<name>X)
-        return source[index + 3] in ["=", "!"]
```
Comment (mpkorstanje) [inline python/cucumber_expressions/tree_regexp.py:68]: For consistency between implementations it would be good to keep this similar too. It really helps if all the implementations are similar enough that you can reference a language implementation you do know.

Though the syntax for  a named group is Python-specific so it would be good to add a separate case for that and comment on it.

File: python/cucumber_expressions/expression.py (hunk 3 above)
Comment (mpkorstanje) [reply python/cucumber_expressions/expression.py:108]: It might also be worth while to push this into the parser.

**Comment by mpkorstanje** [comment]
> ♻️ Anything particular you want feedback on?
> Approach, public API changes, format of the named args, whether it's acceptable in general!

I wasn't aware that named capture groups in Python differ from other languages. Where  Java, Javascript and Ruby us  `(?<name>.*)` Go and Python use `(?P<name>.*)`.  It would be good to have that clear in the code.

I'm also missing some error handling around the `:` character, which is now a reserved character for parameter names.

**Comment by mpkorstanje** [comment]
I'm not sure about the release strategy yet. I don't quite have time to sponsor a Java implementation, I'm currently working using the message format everywhere and technical debt that is pulling to the surface.

It does make me favor feature toggles though.

File: python/cucumber_expressions/expression_factory.py (hunk 4 above)
Comment (luke-hill) [reply python/cucumber_expressions/expression_factory.py:25]: Is it worth standardising this check then across all flavours? I have no idea what we do in ruby as I've not dug into this stuff since the initial release some 4/5 years ago

File: python/cucumber_expressions/tree_regexp.py [hunk 7]
```diff
@@ -51,22 +57,30 @@ def create_group_builder(self, regexp):
...
+        return source[index + 1 : index + 3] == "P<" and source[index + 3] != "?"
+
+    @staticmethod
+    def extract_named_group_name(source: str, index: int) -> str:
+        """
+        Extract the name of a named capturing group, e.g., (?P<name>...) returns "name".
+        """
+        group_name_start = index + 3
+        group_name_end = source.find(">", group_name_start)
+        return source[group_name_start:group_name_end]
```
Comment (luke-hill) [inline python/cucumber_expressions/tree_regexp.py:73]: My python is basically non existent, but above we're using `a : b` and here we're using `a:b` - Do they mean diff things if not maybe keep things standard?

File: python/cucumber_expressions/tree_regexp.py [hunk 8]
```diff
@@ -51,22 +57,30 @@ def create_group_builder(self, regexp):
...
-            # (?>X)
-            return True
-        # (?<=X) or (?<!X) else (?<name>X)
-        return source[index + 3] in ["=", "!"]
 
-    @property
-    def group_builder(self):
-        return self._group_builder
+        # If it's a named group (e.g., (?P<name>...)), it's still a capturing group
+        if source[index + 2] == "P" and source[index + 3] == "<":
```
Comment (luke-hill) [inline python/cucumber_expressions/tree_regexp.py:82]: Earlier we use a substring over a range and here we're using 2 diff substring char matches. 

In my head we should probably be using a range in all situations

File: python/tests/test_expression.py [hunk 9]
```diff
@@ -47,11 +44,24 @@ def test_cucumber_expression_matches(self, load_test_yamls: dict):
...
-        assert values[0] == 7
+        assert values[0] == (7, None)
+
+    def test_documents_match_arguments_with_names(self):
+        values = match("I have {cuke_count:int} cuke(s)", "I have 7 cukes")
+        assert values[0] == (7, "cuke_count")
+
+    def test_documents_match_arguments_with_names_and_spaces(self):
+        values = match(
+            "I have {  cuke_count : int  } cuke(s) and {gherkin_count: int} gherkin(s)",
```
Comment (luke-hill) [inline python/tests/test_expression.py:60]: Is this a python specific interpretation. I'm 90% sure we don't permit spaced out arguments inside the braces. but again I'd need to triple check

**Review by luke-hill — COMMENTED** [review]
One thing I wanted to ask is whether you'd want to get both named for param types and regex out simultaneously or whether you'd want / consider doing them separately.

Purely thinking about the polyglot implementation (Unless you're volunteering to write a bunch of other flavours?)

File: python/cucumber_expressions/argument.py (hunk 2 above)
Comment (jsa34) [reply python/cucumber_expressions/argument.py:28]: Good point - I'll have a think - this was mainly done for my benefit when debugging!

Comment (jsa34) [reply python/cucumber_expressions/argument.py:28]: Tuple is a common type used in Python - the type itself should make sense, but I'll review for clarity

File: python/cucumber_expressions/expression_factory.py (hunk 4 above)
Comment (jsa34) [reply python/cucumber_expressions/expression_factory.py:25]: The problem is that I cannot think of an easy way to distinguish a regex from a normal string in Python - the ^$ syntax isn't used, and are generally just strings.

Hence, I thought to try and identify the other way around - seeing if it's a Cucumber Expression.

I was a bit stuck here as I couldn't think of a reliable deterministic manner to distinguish between the two types, so input very welcome!

**Comment by jsa34** [comment]
> > ♻️ Anything particular you want feedback on?
> > Approach, public API changes, format of the named args, whether it's acceptable in general!
> 
> I wasn't aware that named capture groups in Python differ from other languages. Where Java, Javascript and Ruby us `(?<name>.*)` Go and Python use `(?P<name>.*)`. It would be good to have that clear in the code.
> 
> I'm also missing some error handling around the `:` character, which is now a reserved character for parameter names.
> 
> And adding to the shared test set, even if failing would be good too.

Great points  👍. I'll update to reflect these

File: python/tests/test_expression.py (hunk 9 above)
Comment (jsa34) [reply python/tests/test_expression.py:60]: Good question. 

I was considering these like f-strings in python (which are a very similar concept).

See the discussion from ruff (linting and formatting tool) discussing this: https://github.com/astral-sh/ruff/discussions/9785#discussioncomment-8493982

The left and right padding whitespace is suggested as good practice for readability by one of the Python maintainers who responded to the standards the formatting should adhere to.  Happy to not allow any whitespace, though!

File: python/cucumber_expressions/tree_regexp.py (hunk 8 above)
Comment (jsa34) [reply python/cucumber_expressions/tree_regexp.py:82]: I'll take a look 👍

File: python/cucumber_expressions/argument.py [hunk 10]
```diff
@@ -1,42 +1,52 @@
...
+        self.group = group
+        self.parameter_type = parameter_type
+        self.name = name
 
     @staticmethod
     def build(
-        tree_regexp: TreeRegexp, text: str, parameter_types: List
+        tree_regexp: TreeRegexp,
+        text: str,
+        parameter_types_and_names: List[Tuple[ParameterType, Optional[str]]],
```
[the remark this answers is not in the document]
Comment (jsa34) [reply python/cucumber_expressions/argument.py:22]: Good point!

File: python/tests/test_expression.py (hunk 9 above)
Comment (luke-hill) [reply python/tests/test_expression.py:60]: I agree to not permit space on either side of colon. So the "middle bit" should always be `name:capture-expr` However I genuinely don't know about whitespace padding inside braces. I feel we should leave this comment open for people to dig into as/when they have time.

**Comment by luke-hill** [comment]
> > One thing I wanted to ask is whether you'd want to get both named for param types and regex out simultaneously or whether you'd want / consider doing them separately.
> > Purely thinking about the polyglot implementation (Unless you're volunteering to write a bunch of other flavours?)
> 
> Good question! Honest answer: I have no idea. I was just going for consistency but open to suggestions.

I think making `:` an invalid character as a small breaking change feels like a good small incremental change and I don't feel that would be a lot of work. It would also be good because we haven't done big sweeping changes so it gets people prepared for "bigger" changes e.t.c.

I'm also 99% sure it would be almost no barrier for people (I can't think of people using `:` inside a param type name, but I can't be 100% certain obviously.

File: python/cucumber_expressions/argument.py (hunk 2 above)
Comment (neskk) [reply python/cucumber_expressions/argument.py:28]: Syntax: `{param_name:param_type}`
How do you represent the `Optional[str]` when param_type is not present? With `None`?
For simplicity sake, could it assume `string` type if param_type is omitted?
I think this error is a bit redundant because the parser should handle the errors/exceptions. Meaning that if the code is working properly you will never throw this error.

File: python/cucumber_expressions/expression_factory.py (hunk 4 above)
Comment (neskk) [reply python/cucumber_expressions/expression_factory.py:25]: I think that you can still look for the `^`/ `$` indicators.
This is a nice rule that deterministically distinguishes cucumber-expressions from a regexp ones.

> the ^$ syntax isn't used, and are generally just strings.
Pytest-BDD and other python BDD frameworks, **when no specific parser is specified (default)**, they should check if `^`/ `$` are present and if so treat it as a regular expression. If not, treat it as a cucumber-expression.
This type of regexp step-def is different from using `parsers.re`, but I think it could be a replacement for it all together.

File: python/cucumber_expressions/tree_regexp.py (hunk 7 above)
Comment (neskk) [reply python/cucumber_expressions/tree_regexp.py:73]: In the previous expression the spaces around `:` can be removed to keep codestyle consistent.
In python if you have a list, you can slice it with the `[start_pos:end_pos]` operator. It does not care if you have spaces around start_pos/end_pos.

File: python/cucumber_expressions/regular_expression.py [hunk 11]
```diff
@@ -1,19 +1,24 @@
...
-from typing import Optional, List
+from collections.abc import Generator
+from typing import Optional, Union, List
 
 from cucumber_expressions.argument import Argument
 from cucumber_expressions.parameter_type import ParameterType
 from cucumber_expressions.parameter_type_registry import ParameterTypeRegistry
 from cucumber_expressions.tree_regexp import TreeRegexp
 
+NAMED_CAPTURE_GROUP_REGEX = re.compile(r"\?P<([^>]+)>")
```
[the remark this answers is not in the document]
Comment (neskk) [reply python/cucumber_expressions/regular_expression.py:10]: Is it OK to allow white-spaces in the name of the capture group?
E.g.: `?P<one two>` matches the above regexp.

File: python/tests/test_expression.py (hunk 9 above)
Comment (neskk) [reply python/tests/test_expression.py:60]: Python f-strings can use complex expressions inside `{ }`, effectively being able to execute code logic.

If cucumber/gherkin syntax is not strict about this, I think it's a personal preference to use these extra white-spaces.
I personally dislike having the extra spaces, because my main focus is usually avoiding line-breaks, and these extra spaces can add up.
Making it strict and disallowing spaces could make the code easier to maintain, but at the same time I imagine it's not difficult to optionally allow white-spaces after `{` and before `}`.
For me, disallowing the spaces before and after `:` is also a bit arbitrary - the readability argument still stands and it's feasible.
With age and experience, I tend to prefer opinionated and strict syntaxes, that give you one clear way to do things right, rather than allowing multiple styles and formats.

**Comment by neskk** [comment]
> I'm also 99% sure it would be almost no barrier for people (I can't think of people using `:` inside a param type name, but I can't be 100% certain obviously.

I don't know any programming language that allows `:` in variable/method/class names.
Even allowing spaces in the param type name seems weird to me.

---

I'm coming from the cucumber-vscode-extension language-service repository because I wanted to make it support [pytest-bdd parsers syntax](https://github.com/cucumber/language-service/pull/236).
I actually managed to get it to recognize the syntax, but now it can't distinguish between `{xpto}` being a param type or a named param.
Adding the `{param_name:param_type}` syntax to cucumber expressions, might make it easier for all these tools to integrate and work well in Python.

**Comment by luke-hill** [comment]
Remember that parameter type name constructs come from a string assignment. The ruby example here takes an input set of parameters as keywords then assigns them to the various properties of the parameter type

```
        type = options[:type] || Object
        use_for_snippets = if_nil(options[:use_for_snippets], true)
        prefer_for_regexp_match = if_nil(options[:prefer_for_regexp_match], false)

        parameter_type = CucumberExpressions::ParameterType.new(
          options[:name], # HERE
          options[:regexp],
          type,
          options[:transformer],
          use_for_snippets,
          prefer_for_regexp_match
        )
```

**Comment by neskk** [comment]
I forgot Ruby uses `:` prefix to identify symbols. I'm not sure if it collides with the syntax `{param_name:param_type}`

> Remember that parameter type name constructs come from a string assignment. The ruby example here takes an input set of parameters as keywords then assigns them to the various properties of the parameter type

I not sure I understand your point. I'm thinking `param_type_name` is the name of the type of the parameter, and `param_name` is the name of the argument that will be populated on the receiver function. E.g.:
```python
@when("bla bla bla {arg1:param_type_1} on {arg2:param_type_1}")
def bla_bla_bla(arg1: param_type_1, arg2: param_type_1):
  // do something with arg1, arg2
  pass
```

**Comment by neskk** [comment]
I think this is only an issue with Ruby and even then, the `:` must be the first char in a name, it can never be in the middle of the name.
Anyway, I hope we can make progress on this, because the Cucumber linking and autocomplete plugins for python are broken and unusable with anything other than plain text step-defs (no support for parameters).
